src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -119,7 +120,9 @@ internal CommitManager(ITextView textView, RecentItemsManager recentItemsManager
             // Tab, Enter and Null (call invoke commit) are always commit characters. 
             if (typeChar != '\t' && typeChar != '\n' && typeChar != '\0' && !IsCommitCharacter(serviceRules, roslynItem, typeChar, filterText))
             {
-                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+                // Returning None means we complete the current session with a void commit. 
```

Comment:
what's a 'void' commit?

Reply:
Void commit = Commit making no change comparing to already typed by user.

Reply:
How is that different from a "cancel"?

Reply:
Cancel does not kill the current session. None does.

Reply:
Confirming my understanding:

When the editor has encountered what it and the language consider to be a "potential commit character", it lets us know that we should try to commit it.

There are two main exception cases where we do not commit, both of which are due to the chosen completion item having special, non-standard rules about filtering/committing.
1. The chosen item has explicitly declared that the typed character is to be used for _filtering_. That is, the character enters the buffer but the original completion session keeps running.
2. The chosen item has explicitly declared that the typed character is *not* to be used for committing. 

We check the above exception cases in order, so exception 2 actually means that the original completion session should be immediately halted. Then, the problem was that when halted after a `.`, that `.` was then not being considered as a trigger character for the next session.

---

So more to this PR, if `CommitBehavior.Cancel` is returned, that's our way of communicating that we actually know something the editor doesn't, and the completion session should continue on (the case of additional filter characters).

Then, every `CommitBehavior` _except_ `Cancel` means we are, in fact, "doing" the commit. There are three ways to mark it as such, one of which is `CommitBehavior.None`, and the other two non-`Cancel` commit behaviors that just tell the editor whether to handle TypeChar/Tab/Return non-standardly. In all 3 of these cases (including the case of ), the session is completed **and a new session is free to start if a trigger character had been typed**. This is why changing `Cancel` to `None` has the desired effect

---

@ivanbasov Is everything above correct?

Reply:
@dpoeschl , thank you for the summary! It is correct.